### PR TITLE
Fixed, searching based on both the username and name #745

### LINKED ISSUE #745: Search users based on both the username and name
**Is your feature request related to a problem? Please describe.**
Currently, the search function finds the mentor/mentee based on `Name`. Sometimes, users are remembered by their username instead of their name. e.g. @Techno-Disaster :D

**Describe the solution you'd like**
Add a feature such that the users can be searched even by their username.

**Describe alternatives you've considered**
NA

**Additional context**
Refer [this](https://github.com/anitab-org/mentorship-backend/issues/644#issuecomment-661748664) comment for more details.

### PR DESCRIPTION
**Pull request by codesankalp**
Now the search function finds the mentor/mentee based on both username and name.

To solve this issue added a OR operator to search on basis of username as well as name.

Tests: Manually tested by searching for user with the username and name

Fixes #745

checklist : 
- [x] Implemented searching on basis of username.

### CODE REVIEW

**Comment by codesankalp** [comment]
@PrashanthPuneriya review it.

**Review by SanketDG — COMMENTED** [review]
Wondering whether we can do something using a standard [`or_`](https://docs.sqlalchemy.org/en/13/core/sqlelement.html#sqlalchemy.sql.expression.or_)

**Review by PrashanthPuneriya — COMMENTED** [review]
> Wondering whether we can do something using a standard or_

Yes! Implementing within the prevs code logic would be better. Also, I think it is better if we use `AND operation instead of OR operation` because, it would show both the results of `username` and `name` instead of showing the results by `username` only when we don't find any user by `name`. What do you think @SanketDG ?

**Comment by codesankalp** [comment]
@PrashanthPuneriya If we use `AND` operation then it will check for both username and name which results to return an empty list. But according to @SanketDG  idea using `OR` will be good and it's working also (tested by me).
@PrashanthPuneriya @SanketDG please review it.

**Review by SanketDG — DISMISSED** [review]
Looks good, but I just have this feeling that an explcit `or_` is better than `|`

and yes `and` won't work because of obvious reasons :smile:

**Comment by codesankalp** [comment]
So should i make the change with `or_` or `|` this is ok.
In documentation it is mentioned that both have the same functionality.
@SanketDG

**Review by PrashanthPuneriya — DISMISSED** [review]
Good work @codesankalp!
Yes, I was wrong. I got confused with what OR, AND does and thus described OR operation in plain english 😅

**Review by devkapilbansal — APPROVED** [review]
LGTM

**Review by devkapilbansal — DISMISSED** [review]
The changes made in this PR were tested locally. Following are the results:

1. Code review - Done

2. All possible responses (positive and negative tests) were tested as below:

  * _Test1 Description_  Search for user with username 
    _Screenshot/gif_:  
1st

    _Expected Result_ are same as _Actual Result_

  * _Test2 Description_  Search for user with name
    _Screenshot/gif_:  
2nd

     _Expected Result_ are same as _Actual Result_
    
3. Additional Comments: Run `python -m unittest discover tests` to ensure all test cases passed.

4. Status of PR can be Changed to: Ready to Merge.

**Comment by devkapilbansal** [comment]
@vj-codes look in this PR too. This pull request has three approvals already but it has not been provided with any relevant label.

**Comment by vj-codes** [comment]
Review , testing  ✔️ 
Adding label ready to merge

**Review by isabelcosta — CHANGES_REQUESTED** [review]
@codesankalp there are some merge conflicts due to other merges, could you please fix these here. And also, since you would be changing the code, could you please just add a small unit test that tests the feature you just fixed. A simple test that would fail if your wasn't in place.

File: tests/tasks/test_api_list_tasks.py
Comment (codesankalp) [inline tests/tasks/test_api_list_tasks.py]: the tests are not passing due to this and after reformatting the tests passed so the extra two files changed are `tests/tasks/test_api_list_tasks.py` and `tests/tasks/test_api_create_task.py`

Comment (isabelcosta) [inline tests/tasks/test_api_list_tasks.py]:
```suggestion
            f"/mentorship_relation/{self.mentorship_relation_w_admin_user.id}/tasks",
```
can you do it like this? no need to break the string in this case :)

Comment (isabelcosta) [inline tests/tasks/test_api_list_tasks.py]:
```suggestion
            f"/mentorship_relation/{self.mentorship_relation_w_admin_user.id}/tasks",
```

**Review by isabelcosta — CHANGES_REQUESTED** [review]
@codesankalp I suggested a little style change, can you please address it. I can approve after. Thank you for creating a new test 👏

File: tests/tasks/test_api_list_tasks.py
Comment (isabelcosta) [reply tests/tasks/test_api_list_tasks.py]: please replicate this change in all of the lines you changed to be like this

**Review by isabelcosta — APPROVED** [review]
💯 👏  Thank you for addressing the code review so well :)

**Review by devkapilbansal — APPROVED** [review]

**Review by PrashanthPuneriya — APPROVED** [review]

**Comment by isabelcosta** [comment]
I will merge this PR since it was already tested here https://github.com/anitab-org/mentorship-backend/pull/861#pullrequestreview-519025127
Thank you so much for testing this @devkapilbansal 🙌 
Thank you for your contribution @codesankalp 👏

**Pull request merged**
